src/Components/Components/src/Auth/IAuthenticationStateProvider.cs

```diff
@@ -0,0 +1,33 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Components
+{
+    /// <summary>
+    /// Provides information about the authentication state of the current user.
+    /// </summary>
+    public interface IAuthenticationStateProvider
```

Comment:
This is me showing up (as expected) to ask if we're sure why an interface is better than an abstract class for this. I could see us adding more methods/overloads to this. I always get suspicious about maintainablility when I see a bool method parameter.

So with that out of the way. I'll do what I'm here to do.

Are we sure this wouldn't be better off as an abstract class? It feels more like a class to me.

Reply:
I was wondering whether default interface methods would be the future for avoiding this regular decision point. However since we're not in a position to use them on Mono WebAssembly (`netstandard2.0`) currently, and we can't be sure how soon they will become available, it's probably not the solution for now.

As such, yes, I'll convert this to an ABC.

I was thinking of doing this before because I wanted to add some convenience properties, such as a top-level `bool IsAuthenticated` and `string Name` that got their data from the default identity. It's much more obvious and pleasant to write `AuthState.IsAuthenticated` than `AuthState.User.Identity.IsAuthenticated`.

I won't add the convenience properties just now because I'd like this feature and its usage patterns to bake a bit more first. It's something we can consider later by the time we're making templates and samples.

Reply:
OK, done. One consequence of this is that we had a name clash: we had both this service and an associated component that want to be called `AuthenticationStateProvider`. It's awkward because that is the best name for both of them.

I've resolved this by renaming the component to `<CascadingAuthenticationState>`. In a way that is a bit more explicit.

However, I don't really like the fact that `<CascadingAuthenticationState>` is going to show up in people's code at all. It just feels like boilerplate in the template. I'm considering having some way of including it by default even without it being in app code, since if you're not using auth, it doesn't do anything anyway. Two ways this could be done:

 * We change the renderer to put an implicit `<CascadingAuthenticationState>` at the root of the app tree. Similarly, for any future things like this, they all get wrapped around each other and the app tree.
 * **(Preferred)** We don't interfere with the render tree (as it has perf and debugging implications), we just amend the cascading parameter resolution logic so that there are some default framework-provided values at the root. It doesn't stop you from overriding them. We would have an `IAuthenticationState` at the root (along with any future built-in intrinsics), for which the framework would obtain the value from whatever `AuthenticationStateProvider` is registered in DI.

I spoke with @javiercn about this and our conclusion was that, for the short term, we do nothing. We don't want to introduce new concepts until we know there are multiple usage scenarios. So all the templates will have a `<CascadingAuthenticationState>` around the router in `App.razor`.

@rynowak I'd be interested in whether you have views on this or can anticipate other places where we would benefit from a concept like this.

Reply:
> default interface methods

This is not a good solution for this problem I think. Default Interface Methods only allow you to add methods, but you still can not add state, so for things that we want to be open for future extension, ABCs are the superior choice as you aren't limited to what things you can add.

Reply:
I would suggest doing option 1, but making it based on DI. So you can register things that you want to be injected as *cascading value providers*.

Reply:
These options are open to us in the future, but I'd like to keep it out of scope for this PR. It would be a whole new feature which we can't really determine the design for at this stage.

Reply:
Agreed. I don't expect any action right now.